### ISSUE TITLE
Parsing DS decimal strings is locale-dependent

### ISSUE THREAD

**Issue by susmth** [dgobbi/vtk-dicom#130]
DICOM data elements with VR type DS contain character strings representing a fixed-point or floating-point number. The format of these strings is fixed by the standard. However, when vtkDICOM converts these strings to their numeric value, it appears to be using functions from the C or C++ standard library which use different formats for numerical strings, depending on the current locale. As a result, getting the floating-point value of a vtkDICOMValue will sometimes fail, in particular when the value contains a decimal point and the system locale uses a different decimal point character than DICOM and the standard "C" locale. 

To illustrate the issue, I attach a program that reads the pixel spacing attribute from a DICOM file and prints its value: First as a string, then after conversion to doubles. It will also enable the native locale for numeric formats, indicated by the environment variable LC_NUMERIC. 
Running this program on my system with German locale yields:
```
LC_NUMERIC=de_DE.UTF-8 
pixel spacing as string: 0.72916668653488\0.72916668653488 
floating-point values: 0, 0 
```
However, when running in the "C" locale the program yields the expected output:
```
LC_NUMERIC=C
pixel spacing as string: 0.72916668653488\0.72916668653488
floating-point values: 0.729167, 0.729167
```
Interestingly, I cannot replicate the issue by creating a DS vtkDICOMValue manually and reading its value back. This is what the test suite is doing and it appears to work regardless of the system locale.

In the examples above I've used the following file as input: 
ftp://medical.nema.org/medical/dicom/DataSets/WG16/Philips/ClassicSingleFrame/Liver/DICOM/IM_0001

Here is the test program:
```
#include <vtkDICOMMetaData.h>
#include <vtkDICOMParser.h>
#include <vtkSmartPointer.h>

#include <iostream>
#include <locale.h>

int main(int argc, char *argv[]) {
	
	if (argc != 2) {
		std::cerr << "usage: " << argv[0] << " dicomfile" << std::endl;
		return 1;
	}

	// use the native locale setting for formatting numbers and print its value
	const char *locale = setlocale(LC_NUMERIC, "");
	std::cout << "LC_NUMERIC=" << locale << std::endl;

	vtkSmartPointer<vtkDICOMMetaData> metadata = vtkSmartPointer<vtkDICOMMetaData>::New();

	vtkSmartPointer<vtkDICOMParser> parser = vtkSmartPointer<vtkDICOMParser>::New();
	parser->SetFileName(argv[1]);
	parser->SetMetaData(metadata);
	parser->Update();

	if (parser->GetErrorCode()) {
		std::cerr << "could not parse " << argv[1] << std::endl;
		return 1;
	}

	const vtkDICOMValue &value = metadata->GetAttributeValue(DC::PixelSpacing);
	if (!value.IsValid()) {
		std::cerr << "this dataset contains no pixel spacing attribute" << std::endl;
		return 1;
	}

	std::cout << "pixel spacing as string: " << value.AsString() << std::endl;

	double spacing[2];
	value.GetValues(spacing, 2);

	std::cout << "floating-point values: " << spacing[0] << ", " << spacing[1] << std::endl;

	return 0;
}
```

**Comment by dgobbi** [comment]
Thanks for the note, this looks like a high-priority issue.

For DS values, the code uses strtod(..) for AsDouble(), and uses sprintf("%.10g", ..) to go in the other direction.  Both of these are locale-dependent, so I'll have to fix the code so that it uses locale-independent functions instead.

When you tested with vtkDICOMValue, it was probably writing "," instead of "." because it uses sprintf() to write the DS value.  That's why it could be read with strtod().

**Comment by susmth** [comment]
If it helps, STL streams can be set up to use a specific locale and with a bit of hacking they can operate on existing char buffers. Here is an example, entirely devoid of error checking, but perhaps useful:

```
class StringBuffer: public std::streambuf {
public:
	StringBuffer(char *s, size_t len) {
		setg(s, s, s+len);
		setp(s, s+len);
	}
};

double readDouble(char *s, size_t len) {
	double d = 0;
	StringBuffer buf(s, len);
	std::istream str(&buf);
	str.imbue(std::locale("C"));
	str >> d;
	return d;
}

void writeDouble(char *s, size_t len, double d) {
	StringBuffer buf(s, len);
	std::ostream str(&buf);
	str.imbue(std::locale("C"));
	str.precision(10);
	str << d;
}
```

**Comment by dgobbi** [comment]
The info was helpful, I've added a basic fix in commit c143790.  I'm going to continue to investigate this for a bit, in the future I might import netlib's conversion code (http://www.netlib.org/fp/dtoa.c) since it will guarantee exactly the same input/output regardless of what compiler and OS are used.

**Comment by dgobbi** [comment]
This will be fixed in today's release (0.8.3).
